wazuh/wazuh
# Add support for AWS Cloudwatch Logs in aws-s3 module

Hello Team, 

This PR closes #5187 

# Description

This PR adds support for `AWS Cloudwatch Logs` to our `AWS-S3` module.

Cloudwatch Logs is an AWS service used to store and access logs files from Amazon instances (EC2 and Fargate), AWS CloudTrail, Route 53, and other, regardless of their source, as a single and consistent flow of events ordered by time.

AWS Cloudwatch Logs are stored using log groups and log streams. Our AWS module needs to be able to access those log groups, extract their contents and finally send them to analysd.

This PR adds a new `AWSCloudwatchLogs` class which inherits from `AWSService` in the `AWS-S3.py` module.

Changes to `wmodules-aws.c`, `wm_aws.c` and `wm_aws.h` were also necessary to be able to parse and provide the new parameters.


# How the new Service works

This new class connects to `AWS Cloudwatch Logs`, gets the log groups provided by parameter and extracts all logs for each log stream in those log groups. Once the logs are collected they are sent to the `Analysisd` daemon so they will be analyzed and alerts will be raised if applicable.

To avoid getting duplicated events we are using the `aws_service` database to store following variables:

* `Token`: The token retreived from boto3 used to point to the next set of items to return. Its received every time we call `getLogEvents`. The token expires after 24 hours. If the token is expired the next call will return every single event from the beginning if we don't use any additional parameters.

* `start_time`:  The timestamp with the lowest value of all fetched events, expressed as the number of milliseconds after Jan 1, 1970 00:00:00 UTC. Its used to filter the events we get from boto3.

* `end_time`: Same usage and attributes as `start_time` but for the timestamp with the largest value. 

Those variables will be stored in table called `cloudwatch_logs` in `aws_service` database using `log_group`, `log_stream` and `region` as primary key.

Every time the module is executed, it will try to retrieve those parameters from the database. If they don't exists it will meant that the combination of `log_group`, `log_stream` and `region` is new and no filter using token, start_time or end_time will be necessary.

If the parameters are already present in the database for that primary key they will be used to ensure only the logs with a timestamp lower than `start_time` or higher than `end_time` will be fetched, while applying `only_logs_after` parameter if present. Any log with a timestamp between `start_time` and `end_time` will be ignored as they will have already been processed in a previous execution.

Finally, after each execution of the module, the values for `token`, `start_time` and `end_time` will be updated in database accordingly.

# Available parameters

This module needs two new parameters that are used exclusively for the AWSCloudwatchLogs service:

* `aws_log_groups`: a log group name or a list of them (comma-sepparated) where the logs streams are stored in AWS.

* `remove_log_streams`: A bool to indicate whether to remove the log streams after their logs are extracted or not.

Here are the complete set of parameters available for this Service:
```
<wodle name="aws-s3"> 
  <disabled> no </disabled> 
  <interval> 5m </interval> 
  <run_on_start> yes </run_on_start> 
  <service type="cloudwatchlogs"> 
    <aws_account_id> account </aws_account_id> 
    <aws_account_alias> alias </aws_account_alias> 
    <access_key> access </access_key> 
    <secret_key> secret </secret_key> 
    <aws_profile> profile </aws_profile> 
    <iam_role_arn> role </iam_role_arn> 
    <aws_log_groups> log_group1,log_group2 </aws_log_groups> 
    <only_logs_after> only_logs_after </only_logs_after> 
    <regions> region1,region2 </regions> 
    <remove_log_streams> no </remove_log_streams> 
   </service> 
 </wodle> 
```

# Error management

### Invalid log group
If the log group provided with `aws_log_groups` is invalid, doesn't exists or the user doesn't have enough privileges to access it, the following message will appear in the `ossec.log`:

```
The specified "invalid_log_group_name" log group does not exist or insufficient privileges to access it
```

Take into account that the execution will continue, so if there is one or more log groups in `aws_log_groups` their logs will be extracted as expected.

### Insufficient privileges when trying to remove a log stream

If the user doesn't have enough privileges to remove a particular log stream from AWS Cloudwatch Logs or it was removed using the console during the execution the following error message will appear in the `ossec.log`:

```
Error trying to remove "log log_stream_name" log stream from "log_group_name" log group.
```

### Invalid parameters

The validation process of the rest of the existing parameters has not been modified, so their behaviour will not altered by this PR.


# Manual test performed

To ensure everything is working as expected many manual tests have been performed, including the following one, having log data stored in AWS Cloudwatch Logs with timestamps for 2020-JUN-10 and 2020-JUN-11:

## Test Case

1. Remove the database
2. Execute the module without `only_logs_after` parameter.
3. Executing the module again but this time using `<only_logs_after>2020-JUN-20<only_logs_after>`
4. Executing the module again but this time using `<only_logs_after>2020-JUN-01<only_logs_after>`
5. Executing the module again but this time using `<only_logs_after>2020-JUN-10<only_logs_after>`

After every single step the database values, alert.json and ossec.log where checked. The module behaved as expected.

Additionaly, `test_aws.py` were executed. Here are the results:

```
========================== test session starts ===========================
platform linux -- Python 3.7.4, pytest-4.5.0, py-1.8.0, pluggy-0.13.0
rootdir: /home/carlos/GIT/wazuh/wodles/aws/tests
plugins: pudb-0.7.0, cov-2.8.1
collected 11 items                                                       

test_aws.py ...........                                            [100%]

======================= 11 passed in 0.24 seconds ========================
```
